File: sdks/typescript/src/hash.ts
/**
 * Synap TypeScript SDK - Hash Manager
 * 
 * Redis-compatible Hash data structure operations
 */

import { SynapClient } from './client';

/**
 * Hash field-value pair
 */
export interface HashField {
  field: string;
  value: string | number;
}

/**
 * Hash operations manager
 */
export class HashManager {
  constructor(private client: SynapClient) {}

  /**
   * Set field in hash
   */
  async set(key: string, field: string, value: string | number): Promise<boolean> {
    const response = await this.client.sendCommand<{ success: boolean }>('hash.set', {
      key,
      field,
      value: String(value),
    });
    return response.success || false;
  }

  /**
   * Get field from hash
   */
  async get(key: string, field: string): Promise<string | null> {
    const response = await this.client.sendCommand<{ value?: string }>('hash.get', {
      key,
      field,
    });
    return response?.value ?? null;
  }

  /**
   * Get all fields and values from hash
   */
  async getAll(key: string): Promise<Record<string, string>> {
    const response = await this.client.sendCommand<{ fields?: Record<string, string> }>('hash.getall', {
      key,
    });
    return response?.fields ?? {};
  }

  /**
   * Delete field from hash
   */
  async del(key: string, field: string): Promise<number> {
    const response = await this.client.sendCommand<{ deleted?: number }>('hash.del', {
      key,
      field,
    });
    return response?.deleted ?? 0;
  }

  /**
   * Check if field exists in hash
   */
  async exists(key: string, field: string): Promise<boolean> {
    const response = await this.client.sendCommand<{ exists?: boolean }>('hash.exists', {
      key,
      field,
    });
    return response?.exists ?? false;
  }

  /**
   * Get all field names in hash
   */
  async keys(key: string): Promise<string[]> {
    const response = await this.client.sendCommand<{ fields?: string[] }>('hash.keys', {
      key,
    });
    return response?.fields ?? [];
  }

  /**
   * Get all values in hash
   */
  async values(key: string): Promise<string[]> {
    const response = await this.client.sendCommand<{ values?: string[] }>('hash.values', {
      key,
    });
    return response?.values ?? [];
  }

  /**
   * Get number of fields in hash
   */
  async len(key: string): Promise<number> {
    const response = await this.client.sendCommand<{ length?: number }>('hash.len', {
      key,
    });
    return response?.length ?? 0;
  }

  /**
   * Set multiple fields in hash
   * 
   * Supports both object format (backward compatible) and array format (Redis-compatible)
   * 
   * @example
   * ```typescript
   * // Object format (backward compatible)
   * await hash.mset('user:1', { name: 'Alice', age: 30 });
   * 
   * // Array format (Redis-compatible)
   * await hash.mset('user:1', [{ field: 'name', value: 'Alice' }, { field: 'age', value: '30' }]);
   * ```
   */
  async mset(key: string, fields: Record<string, string | number> | Array<{ field: string; value: string | number }>): Promise<boolean> {
    let payload: Record<string, any> = { key };
    
    // Check if fields is an array (new format) or object (backward compatible)
    if (Array.isArray(fields)) {
      // Array format: [{"field": "...", "value": "..."}, ...]
      payload.fields = fields.map(f => ({
        field: f.field,
        value: String(f.value)
      }));
    } else {
      // Object format (backward compatible)
      payload.fields = Object.fromEntries(
        Object.entries(fields).map(([k, v]) => [k, String(v)])
      );
    }
    
    const response = await this.client.sendCommand<{ success?: boolean }>('hash.mset', payload);
    return response?.success ?? false;
  }

  /**
   * Get multiple fields from hash
   */
  async mget(key: string, fields: string[]): Promise<Record<string, string | null>> {
    const response = await this.client.sendCommand<{ values?: Record<string, string | null> }>(
      'hash.mget',
      {
        key,
        fields,
      }
    );
    return response?.values ?? {};
  }

  /**
   * Increment field value by integer
   */
  async incrBy(key: string, field: string, increment: number): Promise<number> {
    const response = await this.client.sendCommand<{ value?: number }>('hash.incrby', {
      key,
      field,
      increment,
    });
    return response?.value ?? 0;
  }

  /**
   * Increment field value by float
   */
  async incrByFloat(key: string, field: string, increment: number): Promise<number> {
    const response = await this.client.sendCommand<{ value?: number }>('hash.incrbyfloat', {
      key,
      field,
      increment,
    });
    return response?.value ?? 0;
  }

  /**
   * Set field only if it doesn't exist
   */
  async setNX(key: string, field: string, value: string | number): Promise<boolean> {
    const response = await this.client.sendCommand<{ created?: boolean }>('hash.setnx', {
      key,
      field,
      value: String(value),
    });
    return response?.created ?? false;
  }
}

File: sdks/rust/src/hash.rs
//! Hash data structure operations
use crate::client::SynapClient;
use crate::error::Result;
use serde_json::json;
use std::collections::HashMap;

/// Hash data structure interface (Redis-compatible)
///
/// Hash is a field-value map, ideal for storing objects.
#[derive(Clone)]
pub struct HashManager {
    client: SynapClient,
}

impl HashManager {
    /// Create a new Hash manager interface
    pub(crate) fn new(client: SynapClient) -> Self {
        Self { client }
    }

    /// Set field in hash
    pub async fn set<K, F, V>(&self, key: K, field: F, value: V) -> Result<bool>
    where
        K: AsRef<str>,
        F: AsRef<str>,
        V: ToString,
    {
        let payload = json!({
            "key": key.as_ref(),
            "field": field.as_ref(),
            "value": value.to_string(),
        });

        let response = self.client.send_command("hash.set", payload).await?;
        Ok(response
            .get("success")
            .and_then(|v| v.as_bool())
            .unwrap_or(false))
    }

    /// Get field from hash
    pub async fn get<K, F>(&self, key: K, field: F) -> Result<Option<String>>
    where
        K: AsRef<str>,
        F: AsRef<str>,
    {
        let payload = json!({
            "key": key.as_ref(),
            "field": field.as_ref(),
        });

        let response = self.client.send_command("hash.get", payload).await?;
        Ok(response
            .get("value")
            .and_then(|v| v.as_str())
            .map(String::from))
    }

    /// Get all fields and values from hash
    pub async fn get_all<K>(&self, key: K) -> Result<HashMap<String, String>>
    where
        K: AsRef<str>,
    {
        let payload = json!({"key": key.as_ref()});
        let response = self.client.send_command("hash.getall", payload).await?;

        let fields = response
            .get("fields")
            .and_then(|v| serde_json::from_value(v.clone()).ok())
            .unwrap_or_default();

        Ok(fields)
    }

    /// Delete field from hash
    pub async fn del<K, F>(&self, key: K, field: F) -> Result<i64>
    where
        K: AsRef<str>,
        F: AsRef<str>,
    {
        let payload = json!({
            "key": key.as_ref(),
            "field": field.as_ref(),
        });

        let response = self.client.send_command("hash.del", payload).await?;
        Ok(response
            .get("deleted")
            .and_then(|v| v.as_i64())
            .unwrap_or(0))
    }

    /// Check if field exists in hash
    pub async fn exists<K, F>(&self, key: K, field: F) -> Result<bool>
    where
        K: AsRef<str>,
        F: AsRef<str>,
    {
        let payload = json!({
            "key": key.as_ref(),
            "field": field.as_ref(),
        });

        let response = self.client.send_command("hash.exists", payload).await?;
        Ok(response
            .get("exists")
            .and_then(|v| v.as_bool())
            .unwrap_or(false))
    }

    /// Get all field names in hash
    pub async fn keys<K>(&self, key: K) -> Result<Vec<String>>
    where
        K: AsRef<str>,
    {
        let payload = json!({"key": key.as_ref()});
        let response = self.client.send_command("hash.keys", payload).await?;

        let keys = response
            .get("fields")
            .and_then(|v| serde_json::from_value(v.clone()).ok())
            .unwrap_or_default();

        Ok(keys)
    }

    /// Get all values in hash
    pub async fn values<K>(&self, key: K) -> Result<Vec<String>>
    where
        K: AsRef<str>,
    {
        let payload = json!({"key": key.as_ref()});
        let response = self.client.send_command("hash.values", payload).await?;

        let values = response
            .get("values")
            .and_then(|v| serde_json::from_value(v.clone()).ok())
            .unwrap_or_default();

        Ok(values)
    }

    /// Get number of fields in hash
    pub async fn len<K>(&self, key: K) -> Result<usize>
    where
        K: AsRef<str>,
    {
        let payload = json!({"key": key.as_ref()});
        let response = self.client.send_command("hash.len", payload).await?;
        Ok(response.get("length").and_then(|v| v.as_u64()).unwrap_or(0) as usize)
    }

    /// Set multiple fields in hash
    pub async fn mset<K>(&self, key: K, fields: HashMap<String, String>) -> Result<bool>
    where
        K: AsRef<str>,
    {
        let payload = json!({
            "key": key.as_ref(),
            "fields": fields,
        });

        let response = self.client.send_command("hash.mset", payload).await?;
        Ok(response
            .get("success")
            .and_then(|v| v.as_bool())
            .unwrap_or(false))
    }

    /// Get multiple fields from hash
    pub async fn mget<K>(
        &self,
        key: K,
        fields: Vec<String>,
    ) -> Result<HashMap<String, Option<String>>>
    where
        K: AsRef<str>,
    {
        let payload = json!({
            "key": key.as_ref(),
            "fields": fields,
        });

        let response = self.client.send_command("hash.mget", payload).await?;

        let values = response
            .get("values")
            .and_then(|v| serde_json::from_value(v.clone()).ok())
            .unwrap_or_default();

        Ok(values)
    }

    /// Increment field value by integer
    pub async fn incr_by<K, F>(&self, key: K, field: F, increment: i64) -> Result<i64>
    where
        K: AsRef<str>,
        F: AsRef<str>,
    {
        let payload = json!({
            "key": key.as_ref(),
            "field": field.as_ref(),
            "increment": increment,
        });

        let response = self.client.send_command("hash.incrby", payload).await?;
        Ok(response.get("value").and_then(|v| v.as_i64()).unwrap_or(0))
    }

    /// Increment field value by float
    pub async fn incr_by_float<K, F>(&self, key: K, field: F, increment: f64) -> Result<f64>
    where
        K: AsRef<str>,
        F: AsRef<str>,
    {
        let payload = json!({
            "key": key.as_ref(),
            "field": field.as_ref(),
            "increment": increment,
        });

        let response = self
            .client
            .send_command("hash.incrbyfloat", payload)
            .await?;
        Ok(response
            .get("value")
            .and_then(|v| v.as_f64())
            .unwrap_or(0.0))
    }

    /// Set field only if it doesn't exist
    pub async fn set_nx<K, F, V>(&self, key: K, field: F, value: V) -> Result<bool>
    where
        K: AsRef<str>,
        F: AsRef<str>,
        V: ToString,
    {
        let payload = json!({
            "key": key.as_ref(),
            "field": field.as_ref(),
            "value": value.to_string(),
        });

        let response = self.client.send_command("hash.setnx", payload).await?;
        Ok(response
            .get("created")
            .and_then(|v| v.as_bool())
            .unwrap_or(false))
    }
}
